Firmware/ChameleonMini/Application/MifareUltralight.c

```diff
@@ -485,22 +485,7 @@ uint16_t MifareUltralightAppProcess(uint8_t* Buffer, uint16_t BitCount)
             }
 
             return BitCount;
-        } else if (Cmd == CMD_READ && ByteCount == (CMD_READ_FRAME_SIZE + ISO14443A_CRCA_SIZE) && Buffer[1] == 0) {
-            /* This is a short activation method */
-            State = STATE_ACTIVE;
-            return AppProcess(Buffer, CMD_READ_FRAME_SIZE);
-        } else {
-            /* Unknown command. Enter halt state */
-            State = STATE_IDLE;
-        }
-        break;
-
-    case STATE_READY2:
-        ByteCount = (BitCount + 7) >> 3;
-        if (ISO14443AWakeUp(Buffer, &BitCount, CardATQAValue, FromHalt)) {
-            State = FromHalt ? STATE_HALT : STATE_IDLE;
-            return ISO14443A_APP_NO_RESPONSE;
-        } else if (Cmd == ISO14443A_CMD_SELECT_CL2) {
+        } else if (Cmd == ISO14443A_CMD_SELECT_CL2 && State == STATE_READY2) {
```

Comment:
How about even going...
```
...
if (ISO14443AWakeUp(Buffer, &BitCount, CardATQAValue, FromHalt)) {
    State = FromHalt ? STATE_HALT : STATE_IDLE;
    return ISO14443A_APP_NO_RESPONSE;
} else if (Cmd == ISO14443A_CMD_SELECT_CL1 || Cmd == ISO14443A_CMD_SELECT_CL2) {
    uint8_t UidCLX[ISO14443A_CL_UID_SIZE], * dest, addr, bc, SAKValue;
    if(State == STATE_READY1) {
        /* Load UID CL1 and perform anticollision. Since
        * MF Ultralight use a double-sized UID, the first byte
        * of CL1 has to be the cascade-tag byte. */
        UidCLX[0] = ISO14443A_UID0_CT;
        dest = &UidCLX[1];
        addr = UID_CL1_ADDRESS;
        bc = UID_CL1_SIZE;
        SAKValue = CardSAKValue;
    } else if(State == STATE_READY2) {
        dest = UidCLX;
        addr = UID_CL2_ADDRESS;
        bc = UID_CL2_SIZE;
        SAKValue = SAK_CL2_VALUE;
    }
    MemoryReadBlock(dest, addr, bc);
    if (ISO14443ASelect(Buffer, &BitCount, UidCLX, SAKValue)) {
        State++;
    }
    return BitCount;
} else if (Cmd == CMD_READ && ByteCount == (CMD_READ_FRAME_SIZE + ISO14443A_CRCA_SIZE) && Buffer[1] == 0) {
...
```
Kidding... thanks, it was fine by me anyway, and do not even know if my snippet compiles : )

Reply:
I think in this case better to combine READY1 and READY2 into simple READY state and then check only select commands